packages/mui-system/src/createStyled/createStyled.js

```diff
@@ -6,87 +6,73 @@ import getDisplayName from '@mui/utils/getDisplayName';
-  return (
-    typeof tag === 'string' &&
-    // 96 is one less than the char code
-    // for "a" so this is checking that
-    // it's a lowercase character
-    tag.charCodeAt(0) > 96
-  );
-}
+export const systemDefaultTheme = createTheme();
 
 // Update /system/styled/#api in case if this changes
 export function shouldForwardProp(prop) {
   return prop !== 'ownerState' && prop !== 'theme' && prop !== 'sx' && prop !== 'as';
 }
 
-export const systemDefaultTheme = createTheme();
-
-const lowercaseFirstLetter = (string) => {
-  if (!string) {
-    return string;
-  }
-  return string.charAt(0).toLowerCase() + string.slice(1);
-};
-
-function resolveTheme({ defaultTheme, theme, themeId }) {
-  return isEmpty(theme) ? defaultTheme : theme[themeId] || theme;
+function resolveTheme(themeId, theme, defaultTheme) {
+  return isObjectEmpty(theme) ? defaultTheme : theme[themeId] || theme;
 }
 
 function defaultOverridesResolver(slot) {
   if (!slot) {
     return null;
   }
-  return (props, styles) => styles[slot];
+  return (_props, styles) => styles[slot];
 }
 
-function processStyleArg(callableStyle, { ownerState, ...props }) {
+function processStyleArg(callableStyle, props) {
   const resolvedStylesArg =
-    typeof callableStyle === 'function' ? callableStyle({ ownerState, ...props }) : callableStyle;
+    typeof callableStyle === 'function' ? callableStyle(props) : callableStyle;
 
   if (Array.isArray(resolvedStylesArg)) {
-    return resolvedStylesArg.flatMap((resolvedStyle) =>
-      processStyleArg(resolvedStyle, { ownerState, ...props }),
-    );
+    return resolvedStylesArg.flatMap((style) => processStyleArg(style, props));
   }
 
-  const mergedState = { ...props, ...ownerState, ownerState };
+  if (Array.isArray(resolvedStylesArg?.variants)) {
+    const { variants, ...otherStyles } = resolvedStylesArg;
 
-  if (
-    !!resolvedStylesArg &&
-    typeof resolvedStylesArg === 'object' &&
-    Array.isArray(resolvedStylesArg.variants)
-  ) {
-    const { variants = [], ...otherStyles } = resolvedStylesArg;
     let result = otherStyles;
-    variants.forEach((variant) => {
-      let isMatch = true;
+    let mergedState; // We might not need it, initalized lazily
+
+    /* eslint-disable no-labels */
+    variantLoop: for (let i = 0; i < variants.length; i += 1) {
+      const variant = variants[i];
+
       if (typeof variant.props === 'function') {
-        isMatch = variant.props(mergedState);
+        mergedState ??= { ...props, ...props.ownerState, ownerState: props.ownerState };
+        if (!variant.props(mergedState)) {
+          continue;
+        }
       } else {
-        Object.keys(variant.props).forEach((key) => {
-          if (ownerState?.[key] !== variant.props[key] && props[key] !== variant.props[key]) {
-            isMatch = false;
+
+        for (const key in variant.props) {
+          if (props[key] !== variant.props[key] && props.ownerState?.[key] !== variant.props[key]) {
+            continue variantLoop;
```

Comment:
I've noticed that variants are quite expensive. For each `Button` rendered, each of its variants is iterated to find the matching ones. Been wondering how to remove/reduce that cost, but I haven't found anything. This is essentially replicating the logic that browsers do with classnames (e.g. `.button-small.button-outlined`) but in JS instead of C++, so it's much slower. This cost is also present with PigmenCSS.

Reply:
I was curious to see if CVA approached this differently, the API looks very close https://cva.style/docs/getting-started/variants. But no, it seems to be the same implementation: https://github.com/joe-bell/cva/blob/main/packages/cva/src/index.ts#L177

Reply:
So there's a few thing we could do, but to start with we have a problem with theme allocations. For example:
```js
const Button = styled('button', {
  name: 'Button',
})(({ theme }) => /* this function here */ ({
  color: theme.palette.fg,
  variants: [
    ...
  ],
}));
```

That function there runs on every render for every button, and allocates that huge style object for each of them. It's an insane cost, and not knowing if that theme object is static or dynamic prevents any sort of preprocessing or optimization that could be done to improve variants.

There's a very simple option, it's to add a memoization wrapper around that function to avoid the cost. I've implemented a draft of this approach, it saves us another 10% runtime cost across the board (so 30% total with the other perf PRs). Here is a draft (no breaking changes): https://github.com/romgrk/material-ui/pull/1

Reply:
I've found something but it needs to solve the style objects allocations first, details here: https://github.com/romgrk/material-ui/pull/1